pkg/adsc/delta_test.go

```diff
@@ -495,3 +500,260 @@ LDS/:
+						{
+							Name:     "default/foo1",
+							Resource: constructMcpResourceWithOptions("foo1", "foo1.bar.com", "192.1.1.11", "2"),
+						},
+						{
+							Name:     "default/foo2",
+							Resource: constructMcpResourceWithOptions("foo2", "foo2.bar.com", "192.1.1.22", "1"),
+						},
+						{
+							Name:     "default/foo3",
+							Resource: constructMcpResourceWithOptions("foo3", "foo3.bar.com", "192.1.1.3", ""),
+						},
+					},
+				},
+			},
+			expectedResources: [][]string{
+				{"foo1", "foo1.bar.com", "192.1.1.11"},
+				{"foo2", "foo2.bar.com", "192.1.1.2"},
+				{"foo3", "foo3.bar.com", "192.1.1.3"},
+			},
+		},
+		{
+			desc: "update-delete-and-create-resources",
+			serverResponses: []*discovery.DeltaDiscoveryResponse{
+				{
+					TypeUrl: gvk.ServiceEntry.String(),
+					Resources: []*discovery.Resource{
+						{
+							Name:     "default/foo2",
+							Resource: constructMcpResourceWithOptions("foo2", "foo2.bar.com", "192.1.1.222", "4"),
+						},
+						{
+							Name:     "default/foo4",
+							Resource: constructMcpResourceWithOptions("foo4", "foo4.bar.com", "192.1.1.4", "1"),
+						},
+					},
+					RemovedResources: []string{"default/foo1", "default/foo3"},
+				},
+			},
+			expectedResources: [][]string{
+				{"foo2", "foo2.bar.com", "192.1.1.222"},
+				{"foo4", "foo4.bar.com", "192.1.1.4"},
+			},
+		},
+		{
+			desc: "update-delete-and-create-resources",
+			serverResponses: []*discovery.DeltaDiscoveryResponse{
+				{
+					TypeUrl: gvk.ServiceEntry.String(),
+					Resources: []*discovery.Resource{
+						{
+							Name:     "default/foo2",
+							Resource: constructMcpResourceWithOptions("foo2", "foo2.bar.com", "192.2.2.22", "3"),
+						},
+						{
+							Name:     "default/foo3",
+							Resource: constructMcpResourceWithOptions("foo3", "foo3.bar.com", "192.1.1.33", ""),
+						},
+					},
+					RemovedResources: []string{"default/foo4"},
+				},
+			},
+			expectedResources: [][]string{
+				{"foo2", "foo2.bar.com", "192.2.2.22"},
+				{"foo3", "foo3.bar.com", "192.1.1.33"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			expectedLastReceived := make(map[string]*discovery.DeltaDiscoveryResponse)
+			for _, response := range tt.serverResponses {
+				expectedLastReceived[response.TypeUrl] = response
+				respCh <- response
+			}
+
+			assert.EventuallyEqual(t, func() bool {
+				client.mutex.Lock()
+				defer client.mutex.Unlock()
+				rec := client.lastReceived
+
+				if rec == nil && len(rec) != len(expectedLastReceived) {
+					return false
+				}
+				for tpe, rsrcs := range expectedLastReceived {
+					if _, ok := rec[tpe]; !ok {
+						return false
+					}
+					if len(rsrcs.Resources) != len(rec[tpe].Resources) {
+						return false
+					}
+				}
+				return true
+			}, true, retry.Timeout(time.Second), retry.Delay(time.Millisecond))
+
+			// wait for the callback updates the store
+			time.Sleep(1 * time.Millisecond)
```

Comment:
This test is flaky, probably from this. Should we use EventuallyEqual for the checks below?

Reply:
ack